From 089c30c117aafaf45e09bf7f860e45a09d859c1a Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Fri, 10 Jun 2022 06:54:17 +0300
Subject: [PATCH] [SW-70527] gaudi2: ras: nic sram rings repair

This patch added support for BIRA, BISR and BIST operations
for NIC SRAM rings.

TODO: Add NIC unit reset after bist to avoid ecc.

Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
Change-Id: Ica786bd7489c004b6040daa2df2b7755f4951217
---

diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index 685e9ac..8f15876 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -39,6 +39,7 @@
 #define DELAY_20_USEC			20
 #define DELAY_104_USEC			104
 #define DELAY_36_USEC			36
+#define DELAY_44_USEC			44
 
 #define GAUDI2_DFX_NUM			5
 #define GAUDI2_MAX_RING_NUM		28
@@ -517,6 +518,75 @@
 
 }
 
+static void gaudi2_nic_ring_pattern(enum gaudi2_sram_self_repair_op op)
+{
+	uint32_t expected_ones, expected_zeros;
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL41, mmPSOC_DFT_APB_WORD0);
+	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL25, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL2, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_8K_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL28, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL3, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_44_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL31, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL4, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_7P5K_USEC);
+	} else {
+		preboot_puts("Invalid sram repair operation\n");
+		return;
+	}
+	regmap_writel(GAUDI2_RING_PATTERN_VAL42, mmPSOC_DFT_APB_WORD0);
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00000000010100000010100000010100;
+	expected_zeros = 0b00000000001011100001011100001011;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00001010000001010000001010000001;
+	expected_zeros = 0b10000101110000101110000101110000;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+	if (op == SRAM_SELF_REPAIR_OP_BIST)
+		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
+	else
+		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00000000000000010000001010000000;
+	expected_zeros = 0b00000000000000101110000101110000;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+
+	if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL41, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+
+		regmap_writel(GAUDI2_RING_PATTERN_VAL8, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+
+		expected_ones = 0b00000000000000000000000000000000;
+		expected_zeros = 0b00000000100100100100100100100100;
+		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+		regmap_writel(GAUDI2_RING_PATTERN_VAL21, mmPSOC_DFT_APB_WORD1);
+
+		expected_ones = 0b00000000000000000000000000000000;
+		expected_zeros = 0b00000000000000000000000000000000;
+		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	}
+}
+
 static void gaudi2_sram_ring_sel(uint32_t dfx_val, uint32_t ring)
 {
 	uint32_t ring_value = BIT(8) | (ring << 3) | 0x5;
@@ -561,7 +631,7 @@
 				[SRAM_RING_TYPE_XBAR_MID] = gaudi2_xbar_mid_ring_pattern,
 				[SRAM_RING_TYPE_VDEC] = gaudi2_vdec_ring_pattern,
 				[SRAM_RING_TYPE_SOB_XDMA] = gaudi2_sob_xdma_ring_pattern,
-				[SRAM_RING_TYPE_NIC] = NULL,
+				[SRAM_RING_TYPE_NIC] = gaudi2_nic_ring_pattern,
 				[SRAM_RING_TYPE_MME] = NULL,
 				[SRAM_RING_TYPE_HMMU_XBAR] = NULL,
 				[SRAM_RING_TYPE_HMMU_2] = NULL,
@@ -584,7 +654,8 @@
 				ring_type == SRAM_RING_TYPE_SOB_ROT ||
 				ring_type == SRAM_RING_TYPE_XBAR_MID ||
 				ring_type == SRAM_RING_TYPE_VDEC ||
-				ring_type == SRAM_RING_TYPE_SOB_XDMA) {
+				ring_type == SRAM_RING_TYPE_SOB_XDMA ||
+				ring_type == SRAM_RING_TYPE_NIC) {
 				gaudi2_sram_id_sel();
 				gaudi2_sram_ring_sel(dfx_value_table[dfx], ring + 1);
 				gaudi2_sram_region_ring_pattern[ring_type](op);
@@ -705,6 +776,7 @@
 
 /* TODO: Add Xbar memory init after bist to avoid ecc. */
 /* TODO: Add vdec memory init after bist to avoid ecc. */
+/* TODO: Add NIC unit reset after bist to avoid ecc. */
 
 static void gaudi2_reset_engines(void)
 {
diff --git a/arch/gaudi2/common/common.h b/arch/gaudi2/common/common.h
index 6ee429f..c6218aa 100644
--- a/arch/gaudi2/common/common.h
+++ b/arch/gaudi2/common/common.h
@@ -263,6 +263,8 @@
 #define GAUDI2_RING_PATTERN_VAL38		0x1000032
 #define GAUDI2_RING_PATTERN_VAL39		0x1000037
 #define GAUDI2_RING_PATTERN_VAL40		0x1000019
+#define GAUDI2_RING_PATTERN_VAL41		0x100004a
+#define GAUDI2_RING_PATTERN_VAL42		0x1000052
 
 
 void gaudi2_axi_drain_ena(void);
